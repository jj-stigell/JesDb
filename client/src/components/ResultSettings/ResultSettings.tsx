import React from 'react'
import { Checkbox } from 'flowbite-react'

interface IResultSettings {
  resultsPerPage: number
  showCompact: boolean
  showFurigana: boolean
  setResultsPerPage: (value: number) => void
  toggleCompactView: () => void
  toggleFurigana: () => void
}

export default function ResultSettings ({
  resultsPerPage,
  showCompact,
  showFurigana,
  setResultsPerPage,
  toggleCompactView,
  toggleFurigana
}: IResultSettings): React.JSX.Element {
  const handleChange = (e: any): void => {
    setResultsPerPage(Number(e.target.value))
  }

  return (
    <div className="flex items-center gap-2" role="group" aria-label="Result settings">
      <div>
        <label
          id="compact_view_label"
          htmlFor="compact_view"
          className="mr-2 text-xs sm:text-base dark:text-white"
        >
          Compact
        </label>
        <Checkbox
          id="compact_view"
          className="mr-4"
          checked={showCompact}
          aria-checked={showCompact}
          aria-labelledby="compact_view_label"
          onChange={toggleCompactView}
        />
      </div>
      <div>
        <label
          id="furigana_label"
          htmlFor="show_furigana"
          className="mr-2 text-xs sm:text-base dark:text-white"
        >
          Furigana
        </label>
        <Checkbox
          id="show_furigana"
          className="mr-4"
          checked={showFurigana}
          aria-checked={showFurigana}
          aria-labelledby="furigana_label"
          onChange={toggleFurigana}
        />
      </div>
      <label
        id="per_page_label"
        htmlFor="per_page_value_selector"
        className="mr-2 text-xs text-gray-900 dark:text-jesdark-200"
      >
        Show results per page
      </label>
      <select
        onChange={handleChange}
        defaultValue={resultsPerPage}
        id="per_page_value_selector"
        aria-labelledby="per_page_label"
        data-testid="per_page_value_selector"
        className={`
          bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg
          focus:ring-blue-500 focus:border-blue-500 p-2 dark:bg-gray-700
          dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
          dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
        `}
      >
        {[5, 10, 25, 50].map((amount: number) => (
          <option key={`perpage_${amount}`} value={amount}>{amount}</option>
        ))}
      </select>
    </div>
  )
}
