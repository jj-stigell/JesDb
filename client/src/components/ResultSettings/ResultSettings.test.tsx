import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import ResultSettings from './ResultSettings'

describe('ResultSettings Component', () => {
  const mockSetResultsPerPage = jest.fn()
  const mockToggleCompactView = jest.fn()
  const mockToggleFurigana = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders with initial prop values', () => {
    const { asFragment } = render(
      <ResultSettings
        resultsPerPage={10}
        showCompact={false}
        showFurigana={true}
        setResultsPerPage={mockSetResultsPerPage}
        toggleCompactView={mockToggleCompactView}
        toggleFurigana={mockToggleFurigana}
      />
    )
    expect(screen.getByLabelText('Compact')).not.toBeChecked()
    expect(screen.getByLabelText('Furigana')).toBeChecked()
    expect(asFragment()).toMatchSnapshot()
  })

  it('calls toggleCompactView when the compact view checkbox is clicked', () => {
    render(
      <ResultSettings
        resultsPerPage={10}
        showCompact={false}
        showFurigana={true}
        setResultsPerPage={mockSetResultsPerPage}
        toggleCompactView={mockToggleCompactView}
        toggleFurigana={mockToggleFurigana}
      />
    )
    fireEvent.click(screen.getByLabelText('Compact'))

    expect(mockToggleCompactView).toHaveBeenCalledTimes(1)
  })

  it('calls toggleFurigana when the show furigana checkbox is clicked', () => {
    render(
      <ResultSettings
        resultsPerPage={10}
        showCompact={false}
        showFurigana={true}
        setResultsPerPage={mockSetResultsPerPage}
        toggleCompactView={mockToggleCompactView}
        toggleFurigana={mockToggleFurigana}
      />
    )
    fireEvent.click(screen.getByLabelText('Furigana'))

    expect(mockToggleFurigana).toHaveBeenCalledTimes(1)
  })

  it('calls setResultsPerPage when a new page number is selected', () => {
    render(
      <ResultSettings
        resultsPerPage={10}
        showCompact={false}
        showFurigana={true}
        setResultsPerPage={mockSetResultsPerPage}
        toggleCompactView={mockToggleCompactView}
        toggleFurigana={mockToggleFurigana}
      />
    )
    fireEvent.change(screen.getByTestId('per_page_value_selector'), { target: { value: 25 } })

    expect(mockSetResultsPerPage).toHaveBeenCalledWith(25)
  })
})
