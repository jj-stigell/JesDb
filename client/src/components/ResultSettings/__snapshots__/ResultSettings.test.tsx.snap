// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultSettings Component renders with initial prop values 1`] = `
<DocumentFragment>
  <div
    aria-label="Result settings"
    class="flex items-center gap-2"
    role="group"
  >
    <div>
      <label
        class="mr-2 text-xs sm:text-base dark:text-white"
        for="compact_view"
        id="compact_view_label"
      >
        Compact
      </label>
      <input
        aria-checked="false"
        aria-labelledby="compact_view_label"
        class="h-4 w-4 rounded border border-gray-300 bg-gray-100 focus:ring-2 dark:border-gray-600 dark:bg-gray-700 text-cyan-600 focus:ring-cyan-600 dark:ring-offset-gray-800 dark:focus:ring-cyan-600 mr-4"
        id="compact_view"
        type="checkbox"
      />
    </div>
    <div>
      <label
        class="mr-2 text-xs sm:text-base dark:text-white"
        for="show_furigana"
        id="furigana_label"
      >
        Furigana
      </label>
      <input
        aria-checked="true"
        aria-labelledby="furigana_label"
        checked=""
        class="h-4 w-4 rounded border border-gray-300 bg-gray-100 focus:ring-2 dark:border-gray-600 dark:bg-gray-700 text-cyan-600 focus:ring-cyan-600 dark:ring-offset-gray-800 dark:focus:ring-cyan-600 mr-4"
        id="show_furigana"
        type="checkbox"
      />
    </div>
    <label
      class="mr-2 text-xs text-gray-900 dark:text-jesdark-200"
      for="per_page_value_selector"
      id="per_page_label"
    >
      Show results per page
    </label>
    <select
      aria-labelledby="per_page_label"
      class="
          bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg
          focus:ring-blue-500 focus:border-blue-500 p-2 dark:bg-gray-700
          dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
          dark:focus:ring-jesdark-500 dark:focus:border-jesdark-500
        "
      data-testid="per_page_value_selector"
      id="per_page_value_selector"
    >
      <option
        value="5"
      >
        5
      </option>
      <option
        selected=""
        value="10"
      >
        10
      </option>
      <option
        value="25"
      >
        25
      </option>
      <option
        value="50"
      >
        50
      </option>
    </select>
  </div>
</DocumentFragment>
`;
